scripts/sqs-push.js

```diff
@@ -33,28 +35,68 @@ class SqsWriteStream extends Writable {
    * @param {Object} queue - An object with a url property
    */
   constructor (queue, options) {
-    super({ objectMode: true })
+    super({
+      objectMode: true,
+      highWaterMark
+    })
     this.queueUrl = queue.url
     this.sqs = new SQS()
+    this.activePromises = new Map()
+    this.decrementActivePromises = this.decrementActivePromises.bind(this)
+    this.paused = false
   }
 
-  async _write (obj, enc, cb) {
-    try {
-      const Entries = obj.map(o => ({ MessageBody: o, Id: uuidv4() }))
-      // TODO: add backpressure, handle memory bloat
-      this.sqs.sendMessageBatch({ Entries, QueueUrl: this.queueUrl }).promise()
-      return cb()
-    } catch (err) {
-      return cb(err)
+  decrementActivePromises (id) {
+    this.activePromises.delete(id)
+    if (this.paused && this.activePromises.size < promiseThreshold / 2) {
+      this.paused = false
+      this.cb()
+    }
+  }
+
+  _write (obj, enc, cb) {
+    if (this.activePromises.size >= promiseThreshold) {
+      this.paused = true
+      this.cb = cb
+    } else {
+      try {
+        const Entries = obj.map((object) => ({
+          MessageBody: object,
+          Id: uuidv4()
+        }))
+        const Id = uuidv4()
+        const promise = this.sqs.sendMessageBatch({
+          Entries,
+          QueueUrl: this.queueUrl
+        })
+          .promise()
+          .then(() => {
+            this.decrementActivePromises(Id)
+          })
+          .catch((error) => {
+            logUpdate(`Error: ${error}`)
+            this.decrementActivePromises(Id)
+          })
```

Comment:
Forgive the intrusion, especially if I'm off base, but is there any reason you're not use `finally` in this case? Instead of duplicating the call to `decrementActivePromises` in the `then` and `catch`, using `finally` would give you this:
```
.promise()
.catch((error) => logUpdate(`Error: ${error}`))
.finally(() => this.decrementActivePromises(id))
```

Reply:
@chuckwondo You can always intrude 😄.  I agree here.  I have become such an `await` zealot that I forgot the proper syntax for `promise.finally()`.  I will update the PR as soon as I sort out @wronk's issue below.

Reply:
@chuckwondo Unfortunately the promise returned by the AWS JS SDK is [ECMAScript 2015](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/using-promises.html) which does not support `finally`.